modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java

```diff
@@ -7,102 +7,193 @@
-     * passed {@link MetadataKey}.
-     * @throws MetadataResolvingException if the instantiation fails.
+     * Resolves the KeyIdObject for a MultiLevel {@link MetadataKeyId}
+     *
+     * @param componentModel model property of the {@link MetadataKeyId} parameter
+     * @param key            key containing the values of each level
+     * @return the KeyIdObject for the {@link MetadataKeyId} parameter
+     * @throws MetadataResolvingException
      */
-    public Object resolve() throws MetadataResolvingException
+    private static Object resolveMultiLevelKey(ComponentModel componentModel, MetadataKey key, Class metadataKeyType) throws MetadataResolvingException
     {
+        final Map<Field, String> fieldValueMap = fillFieldValueMap(metadataKeyType, key);
+
         Object metadataKeyId;
         try
         {
-            metadataKeyId = getType(type).newInstance();
+            metadataKeyId = metadataKeyType.newInstance();
         }
         catch (Exception e)
         {
-            throw new MetadataResolvingException("Could not instantiate metadata key object", UNKNOWN, e);
+            throw buildException(String.format("MetadataKey object of type '%s' from the component '%s' could not be instantiated", metadataKeyType.getSimpleName(), componentModel.getName()), e);
         }
 
         fieldValueMap.entrySet().forEach(entry -> new FieldSetter<Object, String>(entry.getKey()).set(metadataKeyId, entry.getValue()));
         return metadataKeyId;
     }
 
-    private MetadataResolvingException buildException(String message)
+
+    private static Map<Field, String> fillFieldValueMap(Class type, MetadataKey key) throws MetadataResolvingException
+    {
+        final Map<String, Field> metadataKeyParts = getAnnotatedFields(type, MetadataKeyPart.class).stream().collect(Collectors.toMap(Field::getName, Function.identity()));
+        final Map<String, String> currentParts = getCurrentParts(key);
+        final List<String> missingParts = metadataKeyParts.keySet().stream().filter(partName -> !currentParts.containsKey(partName)).collect(Collectors.toList());
+
+        if (!missingParts.isEmpty())
+        {
+            throw new MetadataResolvingException(String.format("The given MetadataKey does not provide all the required levels. Missing levels: %s", missingParts), INVALID_METADATA_KEY);
+        }
+
+        return currentParts.entrySet()
+                .stream()
+                .filter(keyEntry -> metadataKeyParts.containsKey(keyEntry.getKey()))
+                .collect(Collectors.toMap(keyEntry -> metadataKeyParts.get(keyEntry.getKey()), Map.Entry::getValue));
+    }
+
+    private static Map<String, String> getCurrentParts(MetadataKey key) throws MetadataResolvingException
+    {
+        Map<String, String> metadataKeyParts = new HashMap<>();
+        metadataKeyParts.put(key.getPartName(), key.getId());
+
+        while (!key.getChilds().isEmpty())
```

Comment:
We don't want to lie to the user, if they give us an invalid MetadataKey we need to notify him